// @aws-cdk/cloud-assembly-api/lib/cloud-artifact.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudArtifact = void 0;
const fs = require("fs");
const path = require("path");
const cxschema = require("@aws-cdk/cloud-assembly-schema");
const metadata_1 = require("./metadata");
const error_1 = require("./private/error");
/**
 * Represents an artifact within a cloud assembly.
 */
class CloudArtifact {
    /**
     * Read the metadata for the given artifact
     *
     * HISTORICAL OR PRIVATE USE ONLY
     *
     * This is publicly exposed as a static function for downstream libraries that
     * don't use the `CloudAssembly`/`CloudArtifact` API, yet still need to read
     * an artifact's metadata.
     *
     * 99% of consumers should just access `artifact.metadata`.
     */
    static readMetadata(assemblyDirectory, x) {
        const ret = {};
        if (x.additionalMetadataFile) {
            Object.assign(ret, JSON.parse(fs.readFileSync(path.join(assemblyDirectory, x.additionalMetadataFile), 'utf-8')));
        }
        for (const [p, entries] of Object.entries(x.metadata ?? {})) {
            if (ret[p]) {
                ret[p].push(...entries);
            }
            else {
                ret[p] = entries;
            }
        }
        return ret;
    }
    /**
     * Returns a subclass of `CloudArtifact` based on the artifact type defined in the artifact manifest.
     *
     * @param assembly - The cloud assembly from which to load the artifact
     * @param id - The artifact ID
     * @param artifact - The artifact manifest
     * @returns the `CloudArtifact` that matches the artifact type or `undefined` if it's an artifact type that is unrecognized by this module.
     */
    static fromManifest(assembly, id, artifact) {
        // Implementation is defined in a separate file to break cyclic dependencies
        void (assembly), void (id), void (artifact);
        throw new error_1.CloudAssemblyError('Implementation not overridden yet');
    }
    constructor(assembly, id, manifest) {
        this.assembly = assembly;
        this.id = id;
        this.manifest = manifest;
        this.messages = this.renderMessages();
        this._dependencyIDs = manifest.dependencies || [];
    }
    /**
     * Returns the metadata associated with this Cloud Artifact
     */
    get metadata() {
        return CloudArtifact.readMetadata(this.assembly.directory, this.manifest);
    }
    /**
     * Returns all the artifacts that this artifact depends on.
     */
    get dependencies() {
        if (this._deps) {
            return this._deps;
        }
        this._deps = this._dependencyIDs.map(id => {
            const dep = this.assembly.tryGetArtifact(id);
            if (!dep) {
                throw new error_1.CloudAssemblyError(`Artifact ${this.id} depends on non-existing artifact ${id}`);
            }
            return dep;
        });
        return this._deps;
    }
    /**
     * @returns all the metadata entries of a specific type in this artifact.
     */
    findMetadataByType(type) {
        const metadata = this.metadata;
        const result = new Array();
        for (const p of Object.keys(metadata || {})) {
            for (const entry of (metadata || {})[p]) {
                if (entry.type === type) {
                    result.push({ path: p, ...entry });
                }
            }
        }
        return result;
    }
    renderMessages() {
        const messages = new Array();
        for (const [id, metadata] of Object.entries(this.metadata || {})) {
            for (const entry of metadata) {
                let level;
                switch (entry.type) {
                    case cxschema.ArtifactMetadataEntryType.WARN:
                        level = metadata_1.SynthesisMessageLevel.WARNING;
                        break;
                    case cxschema.ArtifactMetadataEntryType.ERROR:
                        level = metadata_1.SynthesisMessageLevel.ERROR;
                        break;
                    case cxschema.ArtifactMetadataEntryType.INFO:
                        level = metadata_1.SynthesisMessageLevel.INFO;
                        break;
                    default:
                        continue;
                }
                messages.push({ level, entry, id });
            }
        }
        return messages;
    }
    /**
     * An identifier that shows where this artifact is located in the tree
     * of nested assemblies, based on their manifests. Defaults to the normal
     * id. Should only be used in user interfaces.
     */
    get hierarchicalId() {
        return this.manifest.displayName ?? this.id;
    }
}
exports.CloudArtifact = CloudArtifact;
